src/main/python/rlbot/agents/hivemind/python_hivemind.py

```diff
@@ -0,0 +1,157 @@
+from typing import Dict
+
+import queue
+import time
+
+from rlbot.botmanager.agent_metadata import AgentMetadata
+from rlbot.botmanager.bot_helper_process import BotHelperProcess
+from rlbot.messages.flat import MatchSettings
+from rlbot.utils.game_state_util import GameState
+from rlbot.utils.logging_utils import get_logger
+from rlbot.utils.rendering.rendering_manager import RenderingManager
+from rlbot.utils.structures.ball_prediction_struct import BallPrediction
+from rlbot.utils.structures.bot_input_struct import PlayerInput
+from rlbot.utils.structures.game_data_struct import GameTickPacket, FieldInfoPacket
+from rlbot.utils.structures.game_interface import GameInterface
+
+
+class PythonHivemind(BotHelperProcess):
+
+    # Terminology:
+    # hivemind - main process controlling the drones.
+    # drone - a bot under the hivemind's control.
+
+    def __init__(self, agent_metadata_queue, quit_event, options):
+        super().__init__(agent_metadata_queue, quit_event, options)
+
+        # Sets up the logger. The string is the name of your hivemind.
+        # This is configured inside your config file under hivemind_name.
+        self.logger = get_logger(options['name'])
+
+        # Give a warning if exec_path was given.
+        if 'exec_path' in options:
+            self.logger.warning(
+                """An exec_path was given, but this is a PythonHivemind, and so it is ignored.
+                Try subclassing SubprocessHivemind if you want to use an executable."""
+            )
+
+        # The game interface is how you get access to things
+        # like ball prediction, the game tick packet, or rendering.
+        self.game_interface = GameInterface(self.logger)
+
+        # drone_indices is a set of bot indices
+        # which requested this hivemind with the same key.
+        self.drone_indices = set()
+
+    def try_receive_agent_metadata(self):
+        """Adds all drones with the correct key to our set of running indices."""
+        while not self.metadata_queue.empty():
+            # Tries to get the next agent from the queue.
+            single_agent_metadata: AgentMetadata = self.metadata_queue.get(timeout=1.0)
+            # Adds the index to the drone_indices.
+            self.drone_indices.add(single_agent_metadata.index)
```

Comment:
I'm concerned that this may not wait long enough for all the metadata to arrive. What if the first bot starts the helper process and sends metadata, then this runs before the second has sent metadata? Then metadata_queue.get() will return the first metadata instantly, and the loop will exit instantly because the queue is now empty. Have you ever seen any behavior like that where not all indices get received?

The scratch manager seems to have a risky implementation also, but it sleeps a whole second before calling this method, then waits for 0.1 seconds each loop for additional metadata to arrive.

Reply:
I have not had a problem with this. I have run up to 8 bots in a hivemind, beyond that I get errors like this:
`INFO:rlbot[process_configuration.py:67 -  configure_processes() ] psutil.NoSuchProcess no process found with pid 17504`

Increasing the timeout did nothing to remedy the problem.

Reply:
OK! We can improve the situation incrementally, I'm OK with pushing this as it currently is since it doesn't impact the class design.